### ISSUE TITLE
Introducing S3 object system in R

### ISSUE THREAD

**Issue by bobaekang** [ICJIA/r-user-group#13]
## Summary
S3 is the most basic object-oriented programming system in R, which supports a flexible ("polymorphic") design and use of functions.

### Object-oriented programming (OOP)
Object-oriented programming, OOP in short, is a popular programming style, or paradigm, that is based on the concept of "object." In short, a computer program written in the OOP style constructs "objects" of certain "classes" with specific properties or "attributes" and manipulates the "state" of these objects using associated "methods."

In the history of computer programming, OOP has proven to be highly effective in modeling complicated systems as well as interactions between such systems. One crude example of OOP would be creating an object for a bank account, with an attribute for its balance and methods for adding to and subtracting from the balance.

#### OOP systems in R
In principle, everything is an object in R--vectors, data frames, fitted models, and even functions. However, the original design of R language, as well as that of its predecessor, the S language, did not focus on implementing sophisticated OOP patterns while being strongly influenced by another paradigm called functional programming.

Accordingly, OOP systems in R look like something of an afterthought--which they indeed are, devised to meet the practical needs of its users. As a result, there are now multiple, not-so-well-interoperable object systems: S3, S4, and RC (as well as R6 that is built on and improves RC). S3 and S4 are named as such because R essentially inherited them from the S language, which added S3 and S4 *after* the language was first created. And from the perspective of OOP, only the RC/R6 system might be considered properly "object-oriented," fully embracing the key concepts and features of the standard OOP paradigm...

(Proponents of S3 and S4, most notably the creator of S language John M. Chambers, argue that these systems implement "functional OOP," distinctive of more popular "encapsulated OOP." However, in my view, such a distinction between two styles of OOP does not seem widely used or accepted.)

### S3 system
S3 is the most elementary and yet highly useful object system in R. S3 supports a flexible programming design where the same function can have multiple "methods" for different object "classes."

#### S3 class
Class in the S3 system is a special attribute of an R object that is a character vector.  A class of an S3 object can be retrieved using `class()`:
```r
class(as.Date("2018-10-10"))
#> [1] "Date"

class(factor(c("a", "b", "c")))
#> [1] "factor"

class(data.frame(a = 1:3, b = c("a", "b", "c")))
#> [1] "data.frame"
```

Please note that, while S3 is the most basic OOP system implemented in R, the base R has object types that predates S3, called "base types" in [*Advanced R*](https://adv-r.hadley.nz/). Unfortunately, `class()` fails to clearly distinguish between base type and S3 class, and relying on `class()` to work with S3 class can be error-prone:

```r
base_obj <- "2018-10-10"   
s3_obj <- as.Date(base_obj)

# class() can be misleading
class(base_obj)
#> [1] "character"

class(s3_obj)
#> [1] "Date"

# in fact, "character" for base_obj is not a proper S3 class
attributes(base_obj)
#> NULL

attributes(s3_obj)
#> $class
#> [1] "Date"

# meanwhile, Date is not really a base type
typeof(base_obj)
#> [1] "character"

typeof(s3_obj)
#> [1] "double"
``` 

When we use `attributes()`, a proper S3 object will show that it has `$class` attribute while a base object will return `NULL`. Alternatively, `attr(x, "class")` will directly return the class attribute of `x`.

Creating an object with S3 class uses `structure()`:
```r
obj <- structure(1, class = c("bar", "foo"))
obj
#> [1] 1
attr(,"class")
#> [1] "bar" "foo"
```

An S3 object can have multiple classes and the order of class values has practical implications, which we will discuss later in the "Subclass and inheritance" section.

S3 class is extremely flexible, and adding a new class to an object is easy:
```r
which_class <- function(x) attr(x, "class")

add_class <- function(x, .class) {
	stopifnot(is.character(.class))
    
    attr(x, "class") <- c(.class, which_class(x))
    x
  }

a <- structure(c(11, 22, 33), class = "age")
b <- structure(c("Chicago", "London", "Paris"), class = "location")

b <- add_class(b, "city")

which_class(a)
#> [1] "age"

which_class(b)
#> [1] "city"     "location"
```

Removing a specific class from an object is just as easy:
```r
remove_class <- function(x, .class) {
	stopifnot(is.character(.class))
    
    attr(x, "class") <- which_class(x)[which_class(x) != .class]
    x
  }

a <- remove_class(a, "age")
b <- remove_class(b, "city")

which_class(a)
#> NULL

which_class(b)
#> "location"
```

With this knowledge, we can create a function that takes advantage of the S3 system, acting differently based on the class of its argument. For example:
```r
make_ninja <- function(x) structure(x, class = "ninja")

return_it <- function(x) {
  if ("ninja" %in% which_class(x)) {
    print("Poof!")
    NULL
   }
  x
}

return_1 <- return_it(1)
return_1
#> [1] 1

return_x <- return_it("x")
return_x
#> [1] "x"

foo <- structure("x", class = "foo")
return_foo <- return_it(foo)
return_foo
#> [1] "x"
#> attr(,"class")
#> [1] "foo"

ninja <- make_ninja("x")
return_ninja <- return_it(ninja)
#> [1] "Poof!"
return_ninja
#>  NULL
```

Here, `return_it()` normally returns the input value. If the input is of the "ninja" class, however, it disappears! 

#### Generic function and method dispatch
R offers a generic function mechanism that can greatly expand what we just saw in the `return_it` example, that is, sharing a single interface to various "methods" for different object classes. In OOP jargon, this sort of behavior is called "polymorphism."

 Let'e create `return_it_gf` using a generic function:
```r
return_it_gf <- function(x) UseMethod("return_it_gf")```
```

At this point, `return_it_gf` does not do anything useful--it just calls `UseMethod()` with the name of the generic function. Without implementing proper methods, calling `return_it_gf` at this point only throws an error:
```r
return_it_gf(foo)
#> Error in UseMethod("return_it_gf") : 
#>   no applicable method for 'return_it_gf' applied to an object of class "foo"
```

Let's create actual methods to use:
```r
return_it_gf.default <- function(x) x
return_it_gf.ninja <- function(x) {
  print("Poof!")
  NULL
}
```

The naming pattern for each method is `generic.class()` `generic.default()` is a special method that is called, or dispatched, when no class-specific method exists for the input object.

The outputs are as expected:
```r
return_x <- return_it_gf("x")
return_x
#> [1] "x"

return_foo <- return_it_gf(foo)
return_foo
#> [1] "x"
#> attr(,"class")
#> [1] "foo"

return_ninja <- return_it_gf(ninja)
#> [1] "Poof!"
return_ninja
#> NULL
```

Adding new methods to a generic function is simple:
```r
return_it_gf.timebomb <- function(x) {
    print("3...")
    Sys.sleep(1)
    print("2...")
    Sys.sleep(1)
    print("1...")
    Sys.sleep(1)
    "BOOM!"
  }

timebomb <- structure("x", class = "timebomb")
return_it_gf(timebomb)
#> [1] "3..."
#> [1] "2..."
#> [1] "1..."
#> [1] "BOOM!"
```

The same extension of `return_it` could have been implemented without using a generic function:
```r
return_it_ifelse <- function(x) {
  if ("ninja" == which_class(x)[1]) {
    print("Poof!")
    NULL
  } else if ("timebomb" == which_class(x)[1]) {
    print("3...")
    Sys.sleep(1)
    print("2...")
    Sys.sleep(1)
    print("1...")
    Sys.sleep(1)
    "BOOM!"
  } else x
}

return_it_switch <- function(x) {
  if (is.null(which_class(x))) return(x)
  switch(
    which_class(x)[1],
    ninja = {
      print("Poof!")
      NULL
    },
    timebomb = {
      print("3...")
      Sys.sleep(1)
      print("2...")
      Sys.sleep(1)
      print("1...")
      Sys.sleep(1)
      "BOOM!"
    },
    x
  )
}
```

While both look like reasonable options, maintaining such functions gets increasingly more complex and burdensome as the number of required class-specific methods grows. In that case, generic functions with S3 class is the most obvious solution. 

But what about base types without S3 class? Can we create a method for them? The answer is yes:
```r
return_it_basetype <- function(x) paste0(class(x), ": ", x)

return_it_gf.character <- return_it_basetype
return_it_gf.numeric <- return_it_basetype
return_it_gf.logical <- return_it_basetype

return_it_gf("a")
#> [1] "character: a"

return_it_gf(1)
#> [1] "numeric: 1"

return_it_gf(TRUE)
#> [1] "logical: TRUE"
```

This is possible because while base objects do not have a proper S3 class, R gives each base type an "implicit class" that works with generic functions. In fact, what `class()` returns for a base object is its implicit class value.

#### Subclass and inheritance
In the standard OOP, classes are often hierarchically implemented. A class can have a subclass, or a "child," which "inherits" from and shares certain properties or attributes with the original, "parent" class.

S3 does not have a formal inheritance mechanism for its class. Instead, S3 relies on the order of class attribute values to implement such a hierarchical class structure. Let's see how this works with another class:
```r
junior_ninja <- add_class(ninja, "junior_ninja")
which_class(junior_ninja)
#> [1] "junior_ninja" "ninja"

inherits(junior_ninja, "ninja")
#> TRUE

inherits(junior_ninja, "junior_ninja")
#> TRUE
```
`junior_ninja` has two class values: "junior_ninja" and "ninja." Here, `inherits()` is used to check whether an input object inherits from a specific class. However, its name can be somewhat confusing because `inherits()` also returns `TRUE` when the tested class is the object's first class rather than one of its "ancestor" classes.

A more straightforward function to check inheritance might look like this:
```r
is_subclass_of <- function(x, what) {
  inherits(x, what, which = TRUE) > 1 
}

is_subclass_of(junior_ninja, "ninja")
#> TRUE

is_subclass_of(junior_ninja, "junior_ninja")
#> FALSE
```

Here, the `which` argument is used to return the index of the class `inherit()` checks. If the index is greater than 1, the tested class is an "ancestor" class. If 0, the object does not have the tested class. Try `?inherits` to read its documentation for details.

Now let's see how the S3 subclass system works with generic functions. We will give the "junior_ninja" class a method for  `return_it_gf`:
```r
return_it_gf.junior_ninja <- function(x) {
  print("Did I succeed...?")
  x
}

return_junior_ninja <- return_it_gf(junior_ninja )         
#> [1] "Did I succeed...?"
return_junior_ninja
#> [1] "x"
#> attr(,"class")
#> [1] "junior_ninja" "ninja"
```

The "junior_ninja" method of `return_it_gf` uses `print()`  to indicate that a "junior_ninja" class object is not confident about its ninja skill yet. Here, unfortunately, `junior_ninja ` fails to disappear and reveals its identity.

An interesting stuff happens when we use `NextMethod()`, which calls the method for the next, i.e. "parent", class if available.
```r
return_it_gf.junior_ninja <- function(x) {
  print("Did I succeed...?")
  NextMethod()
  x
}

return_junior_ninja <- return_it_gf(junior_ninja)         
#> [1] "Did I succeed...?"
#> [1] "Poof!"
return_junior_ninja
#> [1] "x"
#> attr(,"class")
#> [1] "junior_ninja" "ninja"
```

Now the "junior_ninja" method for `return_it_gf()` first executes `print()` and then calls `NextMethod()`, which dispatches the "ninja" method for `return_it_gf()`: "Poof!"

It is telling, however, that the return value is still not `NULL`.  This shows that even though `NextMethod()` returns `NULL`, the "junior_ninja" method continues on to reach the final line of the function body, `x`. Without this final line, of course, `junior_ninja` will finally succeed:
```r
return_it_gf.junior_ninja <- function(x) {
  print("Did I succeed...?")
  NextMethod()
}

return_junior_ninja <- return_it_gf(junior_ninja)         
#> [1] "Did I succeed...?"
#> [1] "Poof!"
return_junior_ninja
#> NULL
```

### Examples
Here I present three real-world examples of the S3 OOP system in R:

* `print()` relies on the S3 methods to "print" objects of different classes in different formats. For instance, `print()`-ing a `Date` object shows the date in the YYYY-MM-DD format instead of revealing its underlying numeric value.
* `summary()` also replies on the S3 system to give different outputs for different classes. For example, using `summary()` on a numeric vector gives numerical descriptive statistics such as minimum, median, mean, and maximum values. On the other hand, using `summary()` on a factor object gives a count for each possible "level."
* `plot()` has S3 methods to generate different types of graphics based on the class of its input object. For example, using `plot()` on an `lm` object will produce diagnostic plots for the input linear model. 

### A caveat
The current writing is only a cursory introduction to R's S3 OOP system and does not seek to be a comprehensive document on either S3 or OOP in general. Please refer to the materials in [Resources](#resources) and other sources for a more proper discussion on those topics.

### TL;DR
* Use S3 class to organize different object types.
* Use generic functions with class-specific methods to create a cleaner function interface. 
* Use subclasses to reduce replication in your code

### Resources
* [*Advanced Object-Oriented Programming in R*](https://www.apress.com/us/book/9781484229187) by Thomas Mailund*
*  ["Object-oriented programming"](https://en.wikipedia.org/wiki/Object-oriented_programming) in Wikipedia
* ["Object-Oriented Programming, Functional Programming and R"](https://arxiv.org/pdf/1409.3531.pdf) by John M. Chambers
* Part III "Object oriented programming" in [*Advanced R* (2e)](https://adv-r.hadley.nz/) by Hadley Wickham
* Part III "Object-Oriented Programming" in [*Extending R*](https://www.crcpress.com/Extending-R/Chambers/p/book/9781498775717) by John M. Chambers
